daemon/ucrednet.go

```diff
@@ -0,0 +1,96 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2015 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package daemon
+
+import (
+	"errors"
+	"fmt"
+	"net"
+	"strconv"
+	"strings"
+	sys "syscall"
+)
+
+var errNoUID = errors.New("no uid found")
+
+const ucrednetNobody = uint32((1 << 32) - 1)
+
+func ucrednetGetUID(remoteAddr string) (uint32, error) {
+	idx := strings.IndexByte(remoteAddr, ':')
+	if idx < 0 {
+		panic("ucrednetGetUid called on non-ucrednet remoteAddr")
+	}
+	if idx < 1 {
+		return ucrednetNobody, errNoUID
+	}
+
+	uid, err := strconv.ParseUint(remoteAddr[:idx], 10, 32)
+	if err != nil {
+		return ucrednetNobody, err
+	}
+
+	return uint32(uid), nil
+}
+
+type ucrednetAddr struct {
+	net.Addr
+	uid string
+}
+
+func (wa *ucrednetAddr) String() string {
+	return fmt.Sprintf("%s:%s", wa.uid, wa.Addr)
```

Comment:
Agreed. To make matters worse the colon is a very common entry in addresses. "0::1" is an IPv6 address for localhost, for example. Unfortunate consequences if it gets through some day. The `uid=#;...` idea feels simple enough and much less likely to break or have side effects like this. JSON would also be alright, but perhaps too big a gun for something that very likely will never go anywhere else.